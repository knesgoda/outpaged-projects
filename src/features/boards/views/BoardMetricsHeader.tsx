import { memo, useMemo } from "react";
import type { ComponentType, SVGProps } from "react";
import { cn } from "@/lib/utils";
import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from "@/components/ui/tooltip";
import {
  calculateBoardMetrics,
  buildBoardMetricDisplays,
  type BoardMetricDisplay,
} from "./metrics";
import type { BoardViewConfiguration, BoardViewRecord } from "./context";
import { Flame, GitBranch, ListTodo, ShieldCheck, Target, TimerReset } from "lucide-react";

const ICON_MAP: Record<string, ComponentType<SVGProps<SVGSVGElement>>> = {
  kanban: ListTodo,
  target: Target,
  clock: TimerReset,
  flame: Flame,
  shield: ShieldCheck,
  git: GitBranch,
};

const toneClassNames: Record<BoardMetricDisplay["tone"], string> = {
  neutral: "bg-muted/60 text-foreground border-border",
  positive: "bg-emerald-50 text-emerald-700 border-emerald-200 dark:bg-emerald-950 dark:text-emerald-200",
  warning: "bg-amber-50 text-amber-700 border-amber-200 dark:bg-amber-950 dark:text-amber-200",
  critical: "bg-destructive/10 text-destructive border-destructive/40",
};

export interface BoardMetricsHeaderProps {
  items: BoardViewRecord[];
  configuration: BoardViewConfiguration;
  className?: string;
}

function MetricChip({ metric }: { metric: BoardMetricDisplay }) {
  const Icon = ICON_MAP[metric.icon] ?? ListTodo;
  return (
    <Tooltip>
      <TooltipTrigger asChild>
        <div
          className={cn(
            "flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition", 
            "hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring", 
            toneClassNames[metric.tone]
          )}
        >
          <div className="flex items-center gap-3">
            <span className="rounded-full bg-background/60 p-1.5">
              <Icon className="h-4 w-4" aria-hidden="true" />
            </span>
            <div className="flex flex-col">
              <span className="text-xs font-medium uppercase tracking-wide text-muted-foreground">
                {metric.label}
              </span>
              <span className="text-base font-semibold leading-tight">{metric.value}</span>
              {metric.changeLabel ? (
                <span className="text-xs text-muted-foreground">{metric.changeLabel}</span>
              ) : null}
            </div>
          </div>
        </div>
      </TooltipTrigger>
      <TooltipContent side="bottom" className="max-w-xs text-sm">
        {metric.tooltip}
      </TooltipContent>
    </Tooltip>
  );
}

export const BoardMetricsHeader = memo(function BoardMetricsHeader({
  items,
  configuration,
  className,
}: BoardMetricsHeaderProps) {
  const metrics = useMemo(() => {
    const summary = calculateBoardMetrics(items, configuration);
    return buildBoardMetricDisplays(summary);
  }, [configuration, items]);

  if (!metrics.length) {
    return null;
  }

  return (
    <TooltipProvider>
      <div className={cn("grid gap-3 md:grid-cols-2 xl:grid-cols-4", className)}>
        {metrics.map((metric) => (
          <MetricChip key={metric.id} metric={metric} />
        ))}
      </div>
    </TooltipProvider>
  );
});

