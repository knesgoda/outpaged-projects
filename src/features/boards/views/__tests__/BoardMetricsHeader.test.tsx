import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { BoardMetricsHeader } from "../BoardMetricsHeader";
import type { BoardViewConfiguration } from "@/types/boards";

const configuration: BoardViewConfiguration = {
  mode: "table",
  filters: {},
  grouping: { primary: null, swimlaneField: null, swimlanes: [] },
  sort: [],
  columnPreferences: { order: [], hidden: [] },
  timeline: null,
  colorRules: [],
};

describe("BoardMetricsHeader", () => {
  it("renders the metrics snapshot layout", () => {
    const { asFragment } = render(
      <BoardMetricsHeader
        items={[
          {
            id: "1",
            status: "in_progress",
            story_points: 5,
            estimated_hours: 8,
            actual_hours: 6,
            start_date: "2024-03-01",
            due_date: "2024-03-10",
            sla_status: "healthy",
          },
          {
            id: "2",
            status: "done",
            story_points: 3,
            estimated_hours: 5,
            actual_hours: 6,
            start_date: "2024-03-02",
            due_date: "2024-03-08",
            completed: true,
          },
        ]}
        configuration={configuration}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
