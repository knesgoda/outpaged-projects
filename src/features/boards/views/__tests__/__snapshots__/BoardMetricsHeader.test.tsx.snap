// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardMetricsHeader renders the metrics snapshot layout 1`] = `
<DocumentFragment>
  <div
    class="grid gap-3 md:grid-cols-2 xl:grid-cols-4"
  >
    <div
      class="flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring bg-muted/60 text-foreground border-border"
      data-state="closed"
    >
      <div
        class="flex items-center gap-3"
      >
        <span
          class="rounded-full bg-background/60 p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-list-todo h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="6"
              rx="1"
              width="6"
              x="3"
              y="5"
            />
            <path
              d="m3 17 2 2 4-4"
            />
            <path
              d="M13 6h8"
            />
            <path
              d="M13 12h8"
            />
            <path
              d="M13 18h8"
            />
          </svg>
        </span>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          >
            Total work
          </span>
          <span
            class="text-base font-semibold leading-tight"
          >
            2
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            1 done Â· 1 overdue
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring bg-muted/60 text-foreground border-border"
      data-state="closed"
    >
      <div
        class="flex items-center gap-3"
      >
        <span
          class="rounded-full bg-background/60 p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-target h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <circle
              cx="12"
              cy="12"
              r="6"
            />
            <circle
              cx="12"
              cy="12"
              r="2"
            />
          </svg>
        </span>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          >
            Points remaining
          </span>
          <span
            class="text-base font-semibold leading-tight"
          >
            5 pts
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            3 pts done
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring bg-muted/60 text-foreground border-border"
      data-state="closed"
    >
      <div
        class="flex items-center gap-3"
      >
        <span
          class="rounded-full bg-background/60 p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-timer-reset h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 2h4"
            />
            <path
              d="M12 14v-4"
            />
            <path
              d="M4 13a8 8 0 0 1 8-7 8 8 0 1 1-5.3 14L4 17.6"
            />
            <path
              d="M9 17H4v5"
            />
          </svg>
        </span>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          >
            Time logged
          </span>
          <span
            class="text-base font-semibold leading-tight"
          >
            12 h
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            13 h planned
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring bg-amber-50 text-amber-700 border-amber-200 dark:bg-amber-950 dark:text-amber-200"
      data-state="closed"
    >
      <div
        class="flex items-center gap-3"
      >
        <span
          class="rounded-full bg-background/60 p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-flame h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.5 14.5A2.5 2.5 0 0 0 11 12c0-1.38-.5-2-1-3-1.072-2.143-.224-4.054 2-6 .5 2.5 2 4.9 4 6.5 2 1.6 3 3.5 3 5.5a7 7 0 1 1-14 0c0-1.153.433-2.294 1-3a2.5 2.5 0 0 0 2.5 2.5z"
            />
          </svg>
        </span>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          >
            Burndown off track
          </span>
          <span
            class="text-base font-semibold leading-tight"
          >
            5 pts left
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            0 days remaining
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex min-w-[180px] flex-1 items-center justify-between rounded-lg border px-4 py-3 text-left shadow-sm transition hover:shadow-md focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring bg-emerald-50 text-emerald-700 border-emerald-200 dark:bg-emerald-950 dark:text-emerald-200"
      data-state="closed"
    >
      <div
        class="flex items-center gap-3"
      >
        <span
          class="rounded-full bg-background/60 p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-shield-check h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 13c0 5-3.5 7.5-7.66 8.95a1 1 0 0 1-.67-.01C7.5 20.5 4 18 4 13V6a1 1 0 0 1 1-1c2 0 4.5-1.2 6.24-2.72a1.17 1.17 0 0 1 1.52 0C14.51 3.81 17 5 19 5a1 1 0 0 1 1 1z"
            />
            <path
              d="m9 12 2 2 4-4"
            />
          </svg>
        </span>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          >
            SLA compliance
          </span>
          <span
            class="text-base font-semibold leading-tight"
          >
            100%
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            0 breached
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
